build.rs

```diff
@@ -42,44 +41,103 @@ fn main() {
+/// (in standard development directories).
+fn is_in_dir<P>(dir: impl AsRef<Path>, mut predicate: P) -> std::io::Result<bool>
+where P: FnMut(&OsString) -> bool {
+    let mut dirs = Vec::new();
+    dirs.push(dir.as_ref().to_owned());
+    while let Some(dir) = dirs.pop() {
+        let entries = match fs::read_dir(dir) {
+            Ok(e) => e,
+            Err(_) => continue };
+        for entry in entries {
+            let entry = entry?;
+            let ty = entry.file_type()?;
+            if ty.is_file() {
+                if predicate(&entry.file_name()) {
+                    return Ok(true)
+                }
+            } else if ty.is_dir() {
+                dirs.push(entry.path())
+            }
+            // Ignore symlinks
+        }
+    }
+    Ok(false)
+}
+
+
+fn main() {
+    // First, we build the SUNDIALS library, with requested modules with CMake
+
+    let mut lib_loc;
+    let mut inc_dir;
+    let mut library_type = "dylib";
     if cfg!(feature = "build_libraries") {
-        let dst = cmake::Config::new("vendor")
-            .define("CMAKE_INSTALL_LIBDIR", "lib")
-            .define("BUILD_STATIC_LIBS", static_libraries)
-            .define("BUILD_SHARED_LIBS", shared_libraries)
-            .define("BUILD_TESTING", "OFF")
-            .define("EXAMPLES_INSTALL", "OFF")
-            .define("EXAMPLES_ENABLE_C", "OFF")
-            .define("BUILD_ARKODE", feature!("arkode"))
-            .define("BUILD_CVODE", feature!("cvode"))
-            .define("BUILD_CVODES", feature!("cvodes"))
-            .define("BUILD_IDA", feature!("ida"))
-            .define("BUILD_IDAS", feature!("idas"))
-            .define("BUILD_KINSOL", feature!("kinsol"))
-            .define("OPENMP_ENABLE", feature!("nvecopenmp"))
-            .define("PTHREAD_ENABLE", feature!("nvecpthreads"))
-            .build();
-        let dst_disp = dst.display();
-        lib_loc = Some(format!("{}/lib", dst_disp));
-        inc_dir = Some(format!("{}/include", dst_disp));
+        (lib_loc, inc_dir, library_type) = build_vendored_sundials();
     } else {
         lib_loc = get_env_var("SUNDIALS_LIBRARY_DIR");
         inc_dir = get_env_var("SUNDIALS_INCLUDE_DIR");
     }
 
+    if lib_loc.is_none() && inc_dir.is_none() {
+        // No path specified, try to detect if the library is present
+        // on the system.
+        #[cfg(target_family = "unix")] {
+            // Try to find Sundials header files.
+            let std_inc = ["/usr/local/include/", "/usr/include/",
+                           "/usr/include/x86_64-linux-gnu",
+                           "/usr/local/Cellar"];
+            let predicate = |s: &OsString| {
+                s == "nvector_serial.h" };
+            let found = std_inc.iter().any(|d| {
+                is_in_dir(d, predicate).unwrap() });
+            if ! found {
+                (lib_loc, inc_dir, library_type) = build_vendored_sundials();
+            }
+        }
```

Comment:
I'm not too happy about this but couldn't find a better way. Possible options would have been:
- pkg-config, but sundials doesn't ship .pc files
- using `ldconfig -p`, but I don't really know what it is doing...

Did you had a practical case where it was necessary? This feels both not general enough and not so needed.

Reply:
How do you decide whether to use the external sundials or use the vendored one then?

IMHO, the best would be to try to compile a minimal program that includes a function from sundials but I did not find a crate that enables that.  Do you know one?  I'm a bit surprised that none seem to exist since it is a common task for `-sys` crates.

Reply:
I [asked the question on the forum](https://users.rust-lang.org/t/detecting-the-presence-of-c-libraries-sys-crate/86405) but it seems we will have to roll our own procedure.  It seems to me that [openssl-sys](https://github.com/sfackler/rust-openssl/blob/master/openssl-sys/build/find_normal.rs) and [curl-sys](https://github.com/alexcrichton/curl-rust/blob/main/curl-sys/build.rs) (and likely many more) use knowledge that would be better abstracted in an helper library...

Reply:
I changed completely the approach: I try to generate the system bindings and, if they fail, I use the vendored version.